core/job/service/job_service.go

```diff
@@ -322,20 +327,38 @@ func (j JobService) Validate(ctx context.Context, jobTenant tenant.Tenant, jobSp
 	if err != nil {
 		return err
 	}
-	jobMap := make(map[job.Name]*job.Job)
-	destinationToJobsMap := make(map[job.ResourceURN][]*job.Job)
-	for _, jobEntity := range jobsInProject {
-		jobMap[jobEntity.Spec().Name()] = jobEntity
-		if _, ok := destinationToJobsMap[jobEntity.Destination()]; !ok {
-			destinationToJobsMap[jobEntity.Destination()] = []*job.Job{}
+	jobsInProjectWithUpstreams, err := j.upstreamResolver.BulkResolveInternal(ctx, jobTenant.ProjectName(), jobsInProject)
```

Comment:
how about we also use `GetUnresolved` for these jobs that came from db? 
CMIIW I believe we don't really need to resolve the jobs and thus we can avoid introduce `BulkResolveInternal`

Reply:
agree, just examine the code and yeah, use unresolved should be enough

Reply:
move Job with unresolved upstream to the job method